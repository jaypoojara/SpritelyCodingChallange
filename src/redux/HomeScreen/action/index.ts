import {createAsyncThunk} from '@reduxjs/toolkit';
import API from '../../../api';
import axiosInstance from '../../../axios/axios';
import {HomeScreenData} from '../../../types/redux/homeSlice';

export const homeScreenApiAsync = createAsyncThunk<HomeScreenData, undefined>(
  'home/apiCall',
  async () => {
    const response = await axiosInstance.get(API.homeScreen);
    return response.data;
  },
);
