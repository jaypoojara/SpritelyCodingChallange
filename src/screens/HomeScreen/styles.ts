import {StyleSheet} from 'react-native';
import colors from '../../utils/colors';

const styles = StyleSheet.create({
  linearGradientStyle: {flex: 1},
  topViewStyle: {
    marginHorizontal: 30,
  },
  notificationCountText: {
    fontSize: 11,
    fontWeight: 'bold',
    color: colors.white,
  },
  notificationCountView: {
    position: 'absolute',
    height: 17,
    width: 17,
    backgroundColor: colors.darkRed,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
    right: 0,
  },
  weatherInfoView: {
    flexDirection: 'row',
    marginBottom: 25,
    alignItems: 'center',
  },
  weatherInfoText: {
    color: colors.white,
    fontSize: 17,
    fontWeight: '500',
    letterSpacing: 0.7,
    marginLeft: 10,
  },
  unreadNotificationButton: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 12,
    backgroundColor: colors.white,
    borderRadius: 8,
    paddingLeft: 15,
    marginBottom: 20,
  },
  unreadNotificationText: {
    flex: 1,
    paddingHorizontal: 10,
    fontSize: 16,
    fontWeight: '500',
    letterSpacing: 0.4,
  },
  unreadNotificationCloseIcon: {
    paddingHorizontal: 12,
  },
  titleMessageText: {
    marginVertical: 10,
    color: colors.white,
    fontWeight: 'bold',
    fontSize: 23,
    letterSpacing: 0.7,
  },
  bodyView: {
    flex: 1,
    backgroundColor: colors.white,
    borderTopLeftRadius: 35,
    borderTopRightRadius: 35,
    elevation: 5,
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0.3,
    shadowRadius: 10,
  },
  flatListStyle: {
    borderTopLeftRadius: 35,
    borderTopRightRadius: 35,
    flexGrow: 1,
    paddingTop: 30,
    paddingHorizontal: 30,
  },
  emptyViewStyle: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyViewText: {
    color: colors.textColor,
    fontSize: 18,
    fontWeight: '500',
    letterSpacing: 0.8,
  },
  loadingMainView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default styles;
