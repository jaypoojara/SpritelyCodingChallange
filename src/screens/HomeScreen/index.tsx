import React, { useEffect } from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  FlatList,
  ActivityIndicator,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import LinearGradient from 'react-native-linear-gradient';
import { homeScreenApiAsync } from '../../redux/HomeScreen/action';
import { AppDispatch, RootState } from '../../redux/store';
import { HomeInitialState } from '../../types/redux/homeSlice';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import Ionicons from 'react-native-vector-icons/Ionicons';
import Header from '../../components/Header';
import Index from '../../components/FeaturesItem';
import Styles from './styles';
import colors from '../../utils/colors';
import ListEmptyComponent from '../../components/EmptyView';
import LoadingIndicator from '../../components/LoadingIndicator';

export default function HomeScreen() {
  const homeReducer = useSelector<RootState>(
    state => state.homeReducer,
  ) as HomeInitialState;

  const dispatch = useDispatch<AppDispatch>();

  useEffect(() => {
    dispatch(homeScreenApiAsync());
  }, [dispatch]);

  useEffect(() => { }, [homeReducer]);

  const getUserCurrentTime = () => {
    const currentHour = new Date().getHours();
    if (currentHour < 12) {
      return 'Good Morning';
    } else if (currentHour < 18) {
      return 'Good Afternoon';
    } else {
      return 'Good Evening';
    }
  };

  const renderContent = () => {
    if (homeReducer.loading) {
      return (
        <View testID='loading-id' style={Styles.loadingMainView}>
          <LoadingIndicator />
        </View>
      );
    }
    return (
      <>
        <View style={Styles.topViewStyle}>
          <Header notificationCount={homeReducer.data?.notficationCount} />

          <Text style={Styles.titleMessageText}>
            {getUserCurrentTime()}, {homeReducer.data?.firstName}
          </Text>

          <View style={Styles.weatherInfoView}>
            <Ionicons name="sunny" size={30} color={colors.darkYellow} />
            <Text style={Styles.weatherInfoText}>
              {homeReducer.data?.weatherHigh || '-'}°C/
              {homeReducer.data?.weatherLow || '-'}°C
            </Text>
          </View>

          {homeReducer.data?.notficationCount ? (
            <TouchableOpacity style={Styles.unreadNotificationButton}>
              <View>
                <MaterialCommunityIcons
                  name="bell"
                  size={30}
                  color={colors.primary}
                />
                <View style={Styles.notificationCountView}>
                  <Text style={Styles.notificationCountText}>
                    {homeReducer.data?.notficationCount}
                  </Text>
                </View>
              </View>
              <Text style={Styles.unreadNotificationText}>
                {homeReducer.data?.notficationCount} unread notifications
              </Text>
              <TouchableOpacity style={Styles.unreadNotificationCloseIcon}>
                <MaterialCommunityIcons
                  name="window-close"
                  size={25}
                  color={colors.primary}
                />
              </TouchableOpacity>
            </TouchableOpacity>
          ) : null}
        </View>

        {/* Body */}
        <View style={Styles.bodyView}>
          <FlatList
            contentContainerStyle={Styles.flatListStyle}
            data={homeReducer.data?.features}
            keyExtractor={item => item}
            renderItem={({ item }) => <Index name={item} />}
            ListEmptyComponent={<ListEmptyComponent />}
          />
        </View>
      </>
    );
  };

  return (
    <LinearGradient
      colors={[colors.primary, colors.secondary]}
      start={{ x: 0.2, y: 0.1 }}
      end={{ x: 0.9, y: 0.4 }}
      style={Styles.linearGradientStyle}>
      {renderContent()}
    </LinearGradient>
  );
}
