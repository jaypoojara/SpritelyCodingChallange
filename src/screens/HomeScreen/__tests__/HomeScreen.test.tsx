import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import HomeScreen from '..';

const mockStore = configureStore([]);

test('renders correctly', () => {
    const initialState = { homeReducer: { error: null, loading: null, data: null, } };
    const store = mockStore(initialState);

    const element = renderer.create(
        <Provider store={store}>
            <HomeScreen />
        </Provider>
    ).toJSON();
    expect(element).toMatchSnapshot();
});

// test('loading view', () => {
//     const initialState = { homeReducer: { error: null, loading: true, data: null, } };
//     const store = mockStore(initialState);
//     render(
//         <Provider store={store}>
//             <HomeScreen />
//         </Provider>
//     );
//     const element = screen.queryByTestId('loading-id');
//     expect(element).toBeTruthy();
// });