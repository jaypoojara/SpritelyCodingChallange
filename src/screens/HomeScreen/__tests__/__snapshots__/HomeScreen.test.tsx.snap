// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<BVLinearGradient
  colors={
    Array [
      4278205031,
      4285898688,
    ]
  }
  endPoint={
    Object {
      "x": 0.9,
      "y": 0.4,
    }
  }
  locations={null}
  startPoint={
    Object {
      "x": 0.2,
      "y": 0.1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginHorizontal": 30,
      }
    }
  >
    <RCTSafeAreaView
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
      testID="home-header-id"
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/asset/icons/spritely_logo.png",
          }
        }
        style={
          Object {
            "resizeMode": "contain",
            "width": 110,
          }
        }
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 25,
            "elevation": 5,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 10,
            "width": 50,
          }
        }
        testID="notification-button-id"
      >
        <View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#003A67",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            󰂚
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontSize": 23,
          "fontWeight": "bold",
          "letterSpacing": 0.7,
          "marginVertical": 10,
        }
      }
    >
      Good Afternoon
      , 
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 25,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#FFCD0C",
              "fontSize": 30,
            },
            undefined,
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 17,
            "fontWeight": "500",
            "letterSpacing": 0.7,
            "marginLeft": 10,
          }
        }
      >
        -
        °C/
        -
        °C
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderTopLeftRadius": 35,
        "borderTopRightRadius": 35,
        "elevation": 5,
        "flex": 1,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 10,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={<ListEmptyComponent />}
      contentContainerStyle={
        Object {
          "borderTopLeftRadius": 35,
          "borderTopRightRadius": 35,
          "flexGrow": 1,
          "paddingHorizontal": 30,
          "paddingTop": 30,
        }
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#484848",
                "fontSize": 18,
                "fontWeight": "500",
                "letterSpacing": 0.8,
              }
            }
          >
            Data not found!
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</BVLinearGradient>
`;
